File: use-go/python/transfers/_progress.py
"""Terminal progress bar for file transfers."""
from __future__ import annotations

import math
import sys
import threading
import time
from typing import IO


class Progress:
    """
    Thread-safe progress reporter that renders a single-line progress bar
    to an output stream (default: stderr) at ~5 Hz.

    Call order:
        p = Progress()
        p.set_manifest(n_files, total_bytes)
        for each file:
            p.start_file(path, size, resume_offset)
            p.add_bytes(chunk_size)  # called many times
            p.finish_file()
        p.stop()
    """

    def __init__(self, out: IO[str] | None = None) -> None:
        self._out = out or sys.stderr
        self._lock = threading.Lock()

        self._total_files = 0
        self._total_bytes = 0
        self._done_files = 0
        self._done_bytes = 0
        self._current_file = ""

        self._start = time.monotonic()
        self._stop_ev = threading.Event()
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    # ── public API ────────────────────────────────────────────────────────────

    def set_manifest(self, total_files: int, total_bytes: int) -> None:
        with self._lock:
            self._total_files = total_files
            self._total_bytes = total_bytes

    def start_file(self, path: str, size: int, offset: int = 0) -> None:
        with self._lock:
            self._current_file = path

    def add_bytes(self, n: int) -> None:
        with self._lock:
            self._done_bytes += n

    def finish_file(self) -> None:
        with self._lock:
            self._done_files += 1

    def stop(self) -> None:
        self._stop_ev.set()
        self._thread.join()
        self._render()                   # final render
        print(file=self._out)            # newline after bar

    # ── internal ──────────────────────────────────────────────────────────────

    def _loop(self) -> None:
        while not self._stop_ev.wait(timeout=0.2):
            self._render()

    def _render(self) -> None:
        with self._lock:
            done_b = self._done_bytes
            total_b = self._total_bytes
            done_f = self._done_files
            total_f = self._total_files
            cur = self._current_file

        elapsed = time.monotonic() - self._start
        speed_mbs = (done_b / elapsed / 1e6) if elapsed > 0 else 0.0

        pct = min(done_b / total_b, 1.0) if total_b > 0 else 0.0

        if speed_mbs > 0 and total_b > done_b:
            eta_s = (total_b - done_b) / 1e6 / speed_mbs
            eta = "ETA " + _fmt_duration(eta_s)
        elif done_b >= total_b > 0:
            eta = "done"
        else:
            eta = "ETA ?"

        bar = _render_bar(pct, 20)
        cur_short = ("…" + cur[-(34):]) if len(cur) > 35 else cur

        line = (
            f"\r{bar} {pct*100:3.0f}% | "
            f"{done_f}/{total_f} files | "
            f"{_fmt_bytes(done_b)}/{_fmt_bytes(total_b)} | "
            f"{speed_mbs:.1f} MB/s | "
            f"{eta} | {cur_short}"
        )
        print(line, end="", file=self._out, flush=True)


def _render_bar(pct: float, width: int) -> str:
    filled = round(pct * width)
    return "[" + "█" * filled + "░" * (width - filled) + "]"


def _fmt_bytes(n: int) -> str:
    for unit in ("B", "KiB", "MiB", "GiB", "TiB"):
        if n < 1024:
            return f"{n:.1f} {unit}" if unit != "B" else f"{n} {unit}"
        n /= 1024
    return f"{n:.1f} PiB"


def _fmt_duration(s: float) -> str:
    s = int(s)
    h, s = divmod(s, 3600)
    m, s = divmod(s, 60)
    if h:
        return f"{h}h{m:02d}m{s:02d}s"
    if m:
        return f"{m}m{s:02d}s"
    return f"{s}s"

File: use-go/client/progress.go
package transfers

import (
	"fmt"
	"io"
	"math"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// Progress tracks transfer progress and renders a live terminal status line.
// It is safe for concurrent use from multiple goroutines.
type Progress struct {
	totalFiles int
	totalBytes int64

	doneFiles atomic.Int64
	doneBytes atomic.Int64

	mu          sync.Mutex
	currentFile string
	currentSize int64

	startTime time.Time
	output    io.Writer

	stopCh chan struct{}
	doneCh chan struct{}
}

// NewProgress creates a Progress that writes to w (typically os.Stderr).
// Call Stop() when the transfer is complete to flush the final line.
func NewProgress(w io.Writer) *Progress {
	p := &Progress{
		output:    w,
		startTime: time.Now(),
		stopCh:    make(chan struct{}),
		doneCh:    make(chan struct{}),
	}
	go p.loop()
	return p
}

// SetManifest initialises totals once the manifest is known.
func (p *Progress) SetManifest(totalFiles int, totalBytes int64) {
	p.totalFiles = totalFiles
	p.totalBytes = totalBytes
}

// StartFile records that a file is now being transferred.
// offset is the resume offset (already-transferred bytes, pre-counted).
func (p *Progress) StartFile(path string, size, offset int64) {
	p.mu.Lock()
	p.currentFile = path
	p.currentSize = size
	p.mu.Unlock()
}

// AddBytes is called with each chunk's byte count as it completes.
func (p *Progress) AddBytes(n int64) {
	p.doneBytes.Add(n)
}

// FinishFile marks the current file as done.
func (p *Progress) FinishFile() {
	p.doneFiles.Add(1)
}

// Stop finalises the progress display and waits for the render goroutine to exit.
func (p *Progress) Stop() {
	close(p.stopCh)
	<-p.doneCh
	fmt.Fprintln(p.output) // leave cursor on a new line
}

// ── render loop ───────────────────────────────────────────────────────────────

func (p *Progress) loop() {
	defer close(p.doneCh)
	tick := time.NewTicker(200 * time.Millisecond)
	defer tick.Stop()
	for {
		select {
		case <-tick.C:
			p.render()
		case <-p.stopCh:
			p.render() // one final render
			return
		}
	}
}

const barWidth = 20

func (p *Progress) render() {
	done := p.doneBytes.Load()
	total := p.totalBytes
	doneF := p.doneFiles.Load()
	totalF := int64(p.totalFiles)

	elapsed := time.Since(p.startTime).Seconds()
	var speedMBs float64
	if elapsed > 0 {
		speedMBs = float64(done) / elapsed / 1e6
	}

	var eta string
	if speedMBs > 0 && total > done {
		remaining := float64(total-done) / 1e6 / speedMBs
		eta = "ETA " + formatDuration(time.Duration(remaining*float64(time.Second)))
	} else if done >= total && total > 0 {
		eta = "done"
	} else {
		eta = "ETA ?"
	}

	pct := 0.0
	if total > 0 {
		pct = math.Min(float64(done)/float64(total), 1.0)
	}

	bar := renderBar(pct, barWidth)

	p.mu.Lock()
	cur := p.currentFile
	p.mu.Unlock()
	if len(cur) > 35 {
		cur = "…" + cur[len(cur)-34:]
	}

	line := fmt.Sprintf("\r%s %3.0f%% | %d/%d files | %s/%s | %.1f MB/s | %s | %s",
		bar,
		pct*100,
		doneF, totalF,
		formatBytes(done), formatBytes(total),
		speedMBs,
		eta,
		cur,
	)
	fmt.Fprint(p.output, line)
}

func renderBar(pct float64, width int) string {
	filled := int(math.Round(pct * float64(width)))
	if filled > width {
		filled = width
	}
	return "[" + strings.Repeat("█", filled) + strings.Repeat("░", width-filled) + "]"
}

func formatBytes(b int64) string {
	const unit = 1024
	if b < unit {
		return fmt.Sprintf("%d B", b)
	}
	div, exp := int64(unit), 0
	for n := b / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %ciB", float64(b)/float64(div), "KMGTPE"[exp])
}

func formatDuration(d time.Duration) string {
	d = d.Round(time.Second)
	h := d / time.Hour
	d -= h * time.Hour
	m := d / time.Minute
	d -= m * time.Minute
	s := d / time.Second
	if h > 0 {
		return fmt.Sprintf("%dh%02dm%02ds", h, m, s)
	}
	if m > 0 {
		return fmt.Sprintf("%dm%02ds", m, s)
	}
	return fmt.Sprintf("%ds", s)
}
